api/v1beta1/conversion.go

```diff
@@ -16,14 +16,160 @@ limitations under the License.
+		Zones:                getZones(in),
+		ControlPlaneEndpoint: in.Spec.ControlPlaneEndpoint,
+	}
+
+	out.Status = CloudStackClusterStatus{
+		FailureDomains: in.Status.FailureDomains,
+		Ready:          in.Status.Ready,
+	}
+	return nil
+}
+
+// getZones maps failure domains to zones
+func getZones(csCluster *v1beta2.CloudStackCluster) []Zone {
+	var zones []Zone
+	for _, failureDomain := range csCluster.Spec.FailureDomains {
+		zone := failureDomain.Zone
+		zones = append(zones, Zone{
+			Name: zone.Name,
+			ID:   zone.ID,
+			Network: Network{
+				Name: zone.Network.Name,
+				ID:   zone.Network.ID,
+				Type: zone.Network.Type,
+			},
+		})
+	}
+	return zones
+}
+
+// GetFailureDomains maps v1beta1 zones to v1beta2 failure domains.
+func GetFailureDomains(csCluster *CloudStackCluster) ([]v1beta2.CloudStackFailureDomainSpec, error) {
+	var failureDomains []v1beta2.CloudStackFailureDomainSpec
+	namespace := csCluster.Namespace
+	for _, zone := range csCluster.Spec.Zones {
+		name, err := GetDefaultFailureDomainName(namespace, csCluster.Name, zone.ID, zone.Name)
+		if err != nil {
+			return nil, err
+		}
+		failureDomains = append(failureDomains, v1beta2.CloudStackFailureDomainSpec{
+			Name: name,
+			Zone: v1beta2.CloudStackZoneSpec{
+				ID:   zone.ID,
+				Name: zone.Name,
+				Network: v1beta2.Network{
+					ID:   zone.Network.ID,
+					Name: zone.Network.Name,
+					Type: zone.Network.Type,
+				},
+			},
+			Domain:  csCluster.Spec.Domain,
+			Account: csCluster.Spec.Account,
+			ACSEndpoint: corev1.SecretReference{
+				Namespace: namespace,
+				Name:      DefaultEndpointCredential,
+			},
+		})
+	}
+	return failureDomains, nil
+}
+
+// GetDefaultFailureDomainName return zoneID as failuredomain name.
+// Default failure domain name is used when migrating an old cluster to a multiple-endpoints supported cluster, that
+// requires to convert each zone to a failure domain.
+// When upgrading cluster using eks-a, a secret named global will be created by eks-a, and it is used by following
+// method to get zoneID by calling cloudstack API.
+// When upgrading cluster using clusterctl directly, zoneID is fetched directly from kubernetes cluster in cloudstackzones.
+func GetDefaultFailureDomainName(namespace string, clusterName string, zoneID string, zoneName string) (string, error) {
+	if len(zoneID) > 0 {
+		return WithZoneID(zoneID, clusterName), nil
+	}
+
+	secret, err := GetK8sSecret(DefaultEndpointCredential, namespace)
+	if err != nil {
+		return "", err
+	}
+
+	// try fetch zoneID using zoneName through cloudstack client
+	zoneID, err = fetchZoneIDUsingCloudStack(secret, zoneName)
+	if err == nil {
+		return WithZoneID(zoneID, clusterName), nil
+	}
+
+	zoneID, err = fetchZoneIDUsingK8s(namespace, zoneName)
+	if err != nil {
+		return "", nil
+	}
+	return WithZoneID(zoneID, clusterName), nil
+}
+
+func WithZoneID(zoneID, clusterName string) string {
+	return utils.WithClusterSuffix(zoneID[len(zoneID)-8:], clusterName)
```

Comment:
this method/feature will go away, right?

Reply:
Yes, removed.